docs/rfc/rfc-105-non-det-process-proposal.md

```diff
@@ -0,0 +1,300 @@
+  (block format, hashes, etc; the same as in ABCI v0.17.0)
+- the validity checks that now the application can perform as part of `ProcessProposal`;
+  i.e., `ProcessProposal` is now part of $valid(v, s)$
+  and may use the application's part of state $s$ in validating $v$.
+
+With the new structure of the implementation of function $valid(v, s)$:
+
+- consensus _agreement_ is not affected and all processes are still required to agree on the same value
+- the consensus _validity_ property is not affected since we are changing the
+  internals of function $valid(v, s)$; consensus _validity_ just requires this function to be true
+
+However, the new structure of the implementation of function $valid(v, s)$
+may affect consensus _termination_, as some implementations of `ProcessProposal` might reject values
+that CometBFT's internal validity checks would otherwise accept.
+In short, $valid(v, s)$ is possibly more restrictive now.
+
+This document focuses on how consensus _termination_ is affected
+by the new structure of function $valid(v, s)$,
+in particular, the different implementations of `ProcessProposal`.
+
+### ABCI 1.0 (and 2.0) Specification
+
+The [ABCI 1.0 specification][abci-spec] imposes a set of new requirements on the application
+so that its implementation of `PrepareProposal` and `ProcessProposal` does not compromise consensus _termination_,
+given the current consensus algorithm implemented in CometBFT (called Tendermint, and described in the [arXiv paper][arxiv]).
+In contrast to $valid(v, s)$, which is defined as a mathematical function used for consensus's formal specification,
+`PrepareProposal` and `ProcessProposal` are understood as _software functions_ (namely, Go function callbacks) in CometBFT.
+We reproduce here the requirements in the ABCI 1.0 (and 2.0) specification that are relevant for this discussion.
+
+Let $p$ and $q$ be two correct processes.
+Let $r_p$ be a round of consensus at height $h$ where $p$ is the proposer.
+Let $s_{p,h-1}$ be $p$'s application's state committed for height $h-1$.
+In other words, $s_{p,h-1}$ is $p$'s view of $s_{h-1}$.
+Let $v_p$ be the block that $p$'s CometBFT passes
+on to the application
+via `RequestPrepareProposal` as proposer of round $r_p$, height $h$,
+known as the _raw proposal_.
+Let $u_p$ be the (possibly modified) block that $p$'s application
+returns via `ResponsePrepareProposal` to CometBFT in round $r_p$, height $h$,
+known as the _prepared proposal_.
+
+* Requirement 3 [`PrepareProposal`, `ProcessProposal`, coherence]: For any two correct processes $p$ and $q$
+  if $q$'s CometBFT calls `RequestProcessProposal` on $u_p$,
+  $q$'s application returns _Accept_ in `ResponseProcessProposal`.
+
+* Requirement 4 [`ProcessProposal`, determinism-1]:
+  `ProcessProposal` is a (deterministic) function of the current
+  state and the block being processed.
+  In other words, for any correct process $p$, and any arbitrary block $u$,
```

Comment:
You're right, I should have use e.g. $r$ to avoid confusing with "$p$ is proposer of round $r_p$" as stated in the notation above. But I'll leave it like that here, as this is taken from the ABCI spec.